ui/components/src/Legend/ListLegend.tsx

```diff
@@ -11,28 +11,110 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import { List } from '@mui/material';
+import { useTheme } from '@mui/material';
+import { VariableSizeList, ListChildComponentProps } from 'react-window';
+import AutoSizer from 'react-virtualized-auto-sizer';
+import { useRef, useEffect, forwardRef } from 'react';
 import { LegendItem } from '../model';
 import { ListLegendItem } from './ListLegendItem';
 
 interface ListLegendProps {
   items: LegendItem[];
 }
 
+// Default height used to start while virtualizing the list before a true
+// height is determined.
+const DEFAULT_ROW_HEIGHT = 24;
+
 /**
- * ListLegend is used when legend.position is 'right' since legend items are stacked
+ * ListLegend is used when legend.position is 'right' since legend items are
+ * stacked. It is also used for `bottom` positioned legends when there are a
+ * large number of items because it is virtualized and easier to visually scan
+ * large numbers of items when there is a single item per row.
  */
 export function ListLegend({ items }: ListLegendProps) {
+  // Storing a ref to the react-window `VariableSizeList`, so we can call
+  // `resetAfterIndex` to resize the list after mouseover/out events to account
+  // for the change in list items on hover.
+  const listRef = useRef<VariableSizeList>(null);
+  // Storing row heights, so we can use dynamic heights, which enables the
+  // user the hover to show the full label, while still having a virtualized
+  // list.
+  const rowHeights = useRef<Record<number, number>>({});
+
+  const theme = useTheme();
+  // Padding value used throughout to adjust the react-window virtual layouts
+  // to simulate padding per the guidance from:
+  // https://github.com/bvaughn/react-window#can-i-add-padding-to-the-top-and-bottom-of-a-list
+  const LIST_PADDING = parseInt(theme.spacing(1), 10);
+
   // show full labels on hover when there are many total series
   const truncateLabels = items.length > 5;
-  return (
-    <List>
-      {items.map((item) => (
+
+  // Gets the row height for a given item to enable the virtualized list to
+  // render the row properly.
+  function getRowHeight(index: number) {
+    const currentHeight = rowHeights.current[index];
+    return currentHeight ?? DEFAULT_ROW_HEIGHT;
+  }
+
+  // Set the height for a given item to enable the virtualized list to
+  // adjust to size changes.
+  function setRowHeight(index: number, size: number) {
+    // Tell the virtualized list that items changed size and need to be
+    // re-evaluated.
+    listRef.current?.resetAfterIndex(0);
+    rowHeights.current = { ...rowHeights.current, [index]: size };
+  }
+
+  // Renderer for virtualized rows in `VariableSizeList`.
+  function ListLegendRow({ index, style }: ListChildComponentProps) {
```

Comment:
I don't normally like nesting component definitions but it's necessary for this virtualized list use case because we need access to some shared values from the parent (`items`, `LIST_PADDING`) that we can't easily pass around because of how the `react-window` apis work.